### DISCUSSION TITLE
python subprocess get nothing

### DISCUSSION THREAD

**Discussion by wanghaisheng** [streamlink/streamlink#3889] [Q&A] [answered]
```
videolink = 'https://www.youtube.com/watch?v=Sg12mmXtUzw'
videoname = 'Sg12mmXtUzw.mp4'
        result = subprocess.Popen(['streamlink  --http-proxy "socks4a://127.0.0.1:1080" --https-proxy "socks5h://127.0.0.1:1080" -o',videoname,videoLink,' best'],stdout=PIPE, stderr=PIPE)
        stdout, stderr = result.communicate()
```
it is weired i can run the cmd in my windows powshell and output the video mp4 file
```
(base) PS D:\Download\audio-visual\youtubechannelAll> streamlink  --http-proxy "socks4a://127.0.0.1:1080" --https-proxy "socks5h://127.0.0.1"  -o GFHeCGcgBbE.mp4  https://www.youtube.com/watch?v=GFHeCGcgBbE best
[cli][info] Found matching plugin youtube for URL https://www.youtube.com/watch?v=GFHeCGcgBbE
[cli][info] Available streams: audio_mp4a, audio_opus, 144p (worst), 240p, 360p, 480p, 720p, 720p60, 1080p60 (best)
[cli][info] Opening stream: 1080p60 (muxed-stream)
[utils.named_pipe][info] Creating pipe streamlinkpipe-59372-1-6296
[utils.named_pipe][info] Creating pipe streamlinkpipe-59372-2-8095
[download][GFHeCGcgBbE.mp4] Written 21.3 MB (7s @ 1.9 MB/s)
```

**Comment by bastimeyer** [comment, marked as answer]
You're not building the `args` argument of `subprocess.Popen` correctly. The sequence you're passing gets turned into a string on Windows and then it gets tokenized according to the rules of the Windows CMD shell. This means you'll have to carefully take a look at every single parameter and also escape special characters accordingly.

Your `args` get turned into the following string, which is nonsense:

> ```
> "streamlink  --http-proxy \"socks4a://127.0.0.1:1080\" --https-proxy \"socks5h://127.0.0.1:1080\" -o" Sg12mmXtUzw.mp4 https://www.youtube.com/watch?v=Sg12mmXtUzw " best"
> ```

- https://github.com/python/cpython/blob/v3.9.6/Lib/subprocess.py#L1359-L1360
- https://github.com/python/cpython/blob/v3.9.6/Lib/subprocess.py#L533-L600

----

https://docs.python.org/3/library/subprocess.html#popen-constructor

> args should be a sequence of program arguments or else a single string or path-like object. By default, the program to execute is the first item in args if args is a sequence. If args is a string, the interpretation is platform-dependent and described below.

> On Windows, if args is a sequence, it will be converted to a string in a manner described in Converting an argument sequence to a string on Windows. This is because the underlying CreateProcess() operates on strings.

https://docs.python.org/3/library/subprocess.html#converting-argument-sequence

----

Btw, you don't have to run Streamlink in a python subprocess. Import Streamlink and use its API instead.

**Reply by wanghaisheng** [reply]
yes  i do want api,but i could not adjust the following snippet into a whole one


```
        print(videoname)
        print(videoLink)


        s = Streamlink()

        s.set_option("http-proxy", 'socks4a://127.0.0.1:1080')

        s.set_option("https-proxy", 'socks5h://127.0.0.1:1080')
        streams = s.streams(videoLink)
        stream = streams['best']

        dappend = []
        fd = stream.open()

        count = 0
        break_after_count = 3

        while True:
            data = fd.read(1024)
            if not data:
                count += 1
                print(count)
                if count == break_after_count:
                    break
            else:
                count = 0
            dappend.append(data)
        url = streams['best'].url
        cap = cv2.VideoCapture(stream)


        fourcc = cv2.VideoWriter_fourcc(*'MP4V')
        out = cv2.VideoWriter(videoname, fourcc, 20.0, (1280,720))

        while(True):
            ret, frame = cap.read()
            out.write(stream.open())
            # cv2.imshow('frame', frame)
            c = cv2.waitKey(1)
            if c & 0xFF == ord('q'):
                break
        cap.release()
        print("END")       
```
